### DISCUSSION TITLE
How to export a boot file function that uses "store"?

### DISCUSSION THREAD

**Discussion by BenJackGill** [quasarframework/quasar#12608] [General - Components / Directives / etc] [answered]
I have a Quasar boot file that checks for a Firebase user, then uploads that user to a Pinia store.

The Promise named `onAuthStateChangedPromise` requires the Quasar boot `store` argument to work, like so: `const userStore = useUserStore(store)`. If I do not pass in the `store` argument I get an error saying `Error: [🍍]: getActivePinia was called with no active Pinia. Did you forget to install pinia?`.

So everything works fine at the moment because I am passing in the `store` argument. But now I want to export the `onAuthStateChangedPromise` promise and use it in other js modules, and I can't move it to the top level and export it because it needs the `store` argument to work.

So how do I export a boot file function that uses the "store" argument?

Here is the boot file in question:

```ts
import { boot } from 'quasar/wrappers';
import { auth } from 'src/config/firebase';
import { onAuthStateChanged } from 'firebase/auth';
import { useUserStore } from 'src/store/user';

export default boot(async ({ router, urlPath, redirect, store }) => {
  // Get the Firebase user
  const onAuthStateChangedPromise = new Promise((resolve, reject) => {
    auth.onAuthStateChanged(
      (firebaseUser) => {
        const userStore = useUserStore(store); // Note I am using "store" here, so I assume onAuthStateChangedPromise must remain inside "boot" and cannot be moved to the top level
        if (firebaseUser) {
          // Set the Pinia state
          userStore.setUser(firebaseUser);
          // Resolve with the Firebase User
          resolve(firebaseUser);
        } else {
          resolve(false);
        }
      },
      (err) => {
        reject(err);
      }
    );
  });
  const onAuthStateInit = () => onAuthStateChangedPromise; // I want to export this function
  const firebaseUser = await onAuthStateInit();
  // Navigate based on user status
  router.beforeEach((to) => {
    to.matched.some((route) => {
      if (
        !firebaseUser &&
        route.meta.requiresAuth &&
        !urlPath.startsWith('/login')
      ) {
        redirect({ name: 'login' });
      }
    });
  });
});
```

**Comment by yusufkandemir** [comment, marked as answer]
Logically that's not possible. But, you are already waiting until the auth gets initialized correctly before initializing your app, so you don't need to wait until that promise gets resolved in any other place. This means you don't need to export or call that function.

Also, you are using `const firebaseUser = await onAuthStateInit();` variable. You are setting its value once, so your route guard will only use that, whether the user logs in, logs out, etc. So, you would also need to fix that. Also, `Array.prototype.some()` is something different, I would suggest looking at [MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) to understand it.

So, if we sum all of these, my suggestion would be to use this:

```js
import { boot } from 'quasar/wrappers';
import { auth } from 'src/config/firebase';
import { onAuthStateChanged } from 'firebase/auth';
import { useUserStore } from 'src/store/user';

export default boot(async ({ router, store }) => {
  const userStore = useUserStore(store);

  // Keep the user state in store updated
  onAuthStateChanged(auth, (firebaseUser) => {
    userStore.setUser(firebaseUser);
  });

  // We await the auth's initialization before moving on
  await new Promise(resolve => {
    const stopObverser = onAuthStateChanged(auth, (firebaseUser) => {
      // as soon as we resolve, the app init will continue and complete
      resolve(firebaseUser);
      // We no longer need this observer, stop it for performance and to avoid any possible problems
      stopObverser();
    });
  });
  
  // At this point, the auth is initialized, and the store has the user set.

  // Navigate based on user status
  router.beforeEach((to, from) => {
    // If auth is not required or we are at the login page, don't do anything
    if (!to.meta.requiresAuth || from.name === 'login') {
      return;
    }

    // If not logged in, redirect to login page
    // Note: I don't know the store structure, but get the user from the store
    // Note: Don't directly save it to a variable to preserve reactivity.
    if (!userStore.user) {
      return { name: 'login };
    }
  });
});
```

**Reply by BenJackGill** [reply]
Thanks this makes a lot more sense now!

**Reply by BenJackGill** [reply]
Is there any particular reason for using `return { name: 'login' };` vs `redirect({ name: 'login' });`? I thought we were supposed to use `redirect()` so I'm surprised `return` also works.

Also should we call `stopObverser()` on the first `auth.onAuthStateChanged`? Or is that that observer supposed to live forever?

**Reply by yusufkandemir** [reply]
Since you are in `router.beforeEach`, which controls the routing, you can just use its features instead of `redirect`. You would use `redirect` if you had another type of logic, i.e. outside `router.beforeEach`. You can also reject the navigation if you want. Here is the docs for navigation guards: https://router.vuejs.org/guide/advanced/navigation-guards.html

About the second point, check out the comment: `Keep the user state in store updated`, so yes, that should live forever to keep the user inside the store updated. So, if the user logs out by regular ways, or maybe due to a failure, or the token expired and couldn't refresh it, whatever that can cause a logout, the observer would set the user to null inside the store. If the user logs out and logs in with another user, it will do the same too. So, the logic is basically this. We are stopping the second observer because it is there just to make sure the auth is initialized before starting up the app, so it only needs to be run once.
